book-content/chapters/06-objects.md

```diff
@@ -0,0 +1,1386 @@
+
+```typescript
+type BaseAwards = "Grammy" | "MercuryPrize" | "Billboard";
+
+type ExtendedAlbumAwards = Record<BaseAwards, boolean> & {
+  [award: string]: boolean;
+};
+
+const extendedNominations: ExtendedAlbumAwards = {
+  Grammy: true,
+  MercuryPrize: false,
+  Billboard: true, // Additional awards can be dynamically added
+  "American Music Awards": true,
+};
+```
+
+This technique would also work when using an interface and the `extends` keyword:
+
+```typescript
+interface BaseAwards {
+  Grammy: boolean;
+  MercuryPrize: boolean;
+  Billboard: boolean;
+}
+
+interface ExtendedAlbumAwards extends BaseAwards {
+  [award: string]: boolean;
+}
+```
+
+This version is preferable because, in general, `interface extends` is preferable to intersections.
+
+Being able to support both default and dynamic keys in our data structures allows a lot of flexibility to adapt to changing requirements in your applications.
+
+### `PropertyKey`
+
+A useful type to know about when working with dynamic keys is `PropertyKey`.
+
+The `PropertyKey` type is a global type that represents the set of all possible keys that can be used on an object, including string, number, and symbol. You can find its type definition inside of TypeScript's ES5 type definitions file:
+
+```typescript
+// inside lib.es5.d.ts
+declare type PropertyKey = string | number | symbol;
+```
+
+Because `PropertyKey` works with all possible keys, it's great for working with dynamic keys where you aren't sure what the type of the key will be.
+
+For example, when using an index signature you could set the key type to `PropertyKey` in order to allow for any valid key type:
+
+```typescript
+type Album = {
+  [key: PropertyKey]: string;
+};
+```
+
+### `object`
+
+Similar to `string`, `number`, and `boolean`, `object` is a global type in TypeScript.
+
+It represents more types than you might expect. Instead of representing only objects like `{}` or `new Object()`, `object` represents any non-primitive type. This includes arrays, functions, and objects.
+
+So a function like this:
+
+```typescript
+function acceptAllNonPrimitives(obj: object) {}
+```
+
+Would accept any non-primitive value:
+
+```typescript
+acceptAllNonPrimitives({});
+acceptAllNonPrimitives([]);
+acceptAllNonPrimitives(() => {});
+```
+
+But error on primitives:
+
+```ts twoslash
+// @errors: 2345
+function acceptAllNonPrimitives(obj: object) {}
+
+// ---cut---
+acceptAllNonPrimitives(1);
+acceptAllNonPrimitives("hello");
+acceptAllNonPrimitives(true);
+```
+
+This means that the `object` type is rarely useful by itself. Using `Record` is usually a better choice. For instance, if you want to accept any object type, you can use `Record<string, unknown>`.
```

Comment:
This is a fallacy - to the best of my knowledge there is no way clear way to create a type that would represent "a plain object". `Record<string, unknown>` has its gotchas too and it just doesn't deliver that: [TS playground](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIBGcUAXMgM5hSgDmA3LgL64LohXIybkZYC8OfERLkARGIA0zRrhgBXEAjDB2yREgAOYAPKEAVhGUUAFOgPkASkfRQAJgB4qNELUnJFAaxDoA7iAA+AEocFlwNCG09Q2MTbnQgxiA)

Reply:
it only kinda works with objects defined `type` because those have "inferrable index signatures" - which is far from obvious